Add Fault Simulation for Dirty Air Filters

rongpengzhang opened the pull request:
## Symptom:

Increased air loop system resistance, resulting in a different system curve. This directly affects the operation of corresponding fans. More specifically, it may lead to an increase in the fan pressure rise, fan energy consumption, as well as the enthalpy of the fan outlet air. It may also lead to a reduction in the airflow rate and thus affects the performance of other system components (e.g., heat transfer performance of heating/cooling coils).
## Model inputs:

Name of the fan affected, name of availability schedule of the fault, name of the fan pressure variation schedule describing the increase of fan pressure rise due to dirty filter, name of the fan curve describing the relationship between fan pressure rise and air flow rate.
## Approach:

New objects will be created for the proposed fault types, namely: 
• FaultModel:Fouling:AirFilters
The fault applies to the following objects:
• Fan:ConstantVolume 
• Fan:OnOff
• Fan:VariableVolume
## ToDo List
- [x] IDD Change
- [x] Code Change
- [x] Test file
- [x] Documentation
- [x] Review (@myoldmopar)

rongpengzhang committed:
New parameters defined for dirty air filter fault

New parameters defined for the dirty air filter fault

3 files
modified:
1) IDD file
2) FaultsManager.cc
3) FaultsManager.hh

rongpengzhang committed:
New parameters defined for dirty air filter fault

rongpengzhang committed:
Faulty Air Filter: IDF read in and check

rongpengzhang committed:
Faulty Air Filter: Fan CAV codes

rongpengzhang committed:
Faulty Air Filter: Fan VAV codes

rongpengzhang committed:
Merge remote-tracking branch 'origin/develop' into 83811490-Fault_DirtyAirFilter

rongpengzhang committed:
Faulty Air Filter: Example file

rongpengzhang commented:
@nrel-bot 
unit tests are being developed.

rongpengzhang committed:
Faulty Air Filter: Example files refinement

rongpengzhang committed:
Faulty Air Filter: Unit Tests

rongpengzhang committed:
Faulty Air Filters: Unit Tests

rongpengzhang committed:
Merge remote-tracking branch 'origin/develop' into 83811490-Fault_DirtyAirFilter

Myoldmopar commented:
@rongpengzhang It looks like you renamed a testfile but didn't change the name in testfiles/CMakeLists.txt, so the CI is reporting an error on that file.  It'll probably come up all green after that fix.

rongpengzhang commented:
@Myoldmopar Thanks for pointing it out! I think that's the reason.

rongpengzhang committed:
Faulty Air Filters: Example File CMakeList

Myoldmopar commented on tst/EnergyPlus/unit/FaultFoulingAirFilters.unit.cc:
```diff
@@ -0,0 +1,137 @@
+    FanDesignFlowRateDec = CalFaultyFanAirFlowReduction( Fan( FanNum ).FanName, Fan( FanNum ).MaxAirFlowRate, Fan( FanNum ).DeltaPress, 
+                           FanFaultyDeltaPressInc * Fan( FanNum ).DeltaPress, CurveNum );
+							
+	EXPECT_GE( FanDesignFlowRateDec, 3.84 ); 
```
@rongpengzhang You can use EXPECT_NEAR and give a tolerance, so like:

`EXPECT_NEAR( 3.845, FanDesignFlowRateDec, 0.005 );` 

if that is close enough.  

As an aside, it is convention (not required) to put the expected value as the first argument in these calls, so that's why I put 3.84 above as the first argument.

rongpengzhang replied:
Thanks, @Myoldmopar ! That's exactly what I needed but did not find.

Myoldmopar replied:
Sure, and you don't have to address all of these immediately.  I'd wait
until I get through the code so you don't have to make a ton of extra
commits.

On Wed, May 20, 2015 at 11:11 AM, Rongpeng Zhang notifications@github.com
wrote:

> In tst/EnergyPlus/unit/FaultFoulingAirFilters.unit.cc
> https://github.com/NREL/EnergyPlus/pull/4894#discussion_r30724938:
> 
> > -   PerfCurve( CurveNum ).Coeff6 = 0.0;
> > -   PerfCurve( CurveNum ).Var1Min = 7.0;
> > -   PerfCurve( CurveNum ).Var1Max = 21.0;
> > -   
> > -   // Inputs: fans
> > -   FanNum = 1;
> > -   Fan( FanNum ).FanName = "Fan_1";
> > -   Fan( FanNum ).FanType = "Fan:VariableVolume";
> > -   Fan( FanNum ).MaxAirFlowRate = 18.194;
> > -   Fan( FanNum ).DeltaPress = 1017.59;
> > -   
> > -   // Run and Check
> > -    FanDesignFlowRateDec = CalFaultyFanAirFlowReduction( Fan( FanNum ).FanName, Fan( FanNum ).MaxAirFlowRate, Fan( FanNum ).DeltaPress,
> > -                           FanFaultyDeltaPressInc \* Fan( FanNum ).DeltaPress, CurveNum );
> > -   
> > -   EXPECT_GE( FanDesignFlowRateDec, 3.84 );
> 
> Thanks, @Myoldmopar https://github.com/Myoldmopar ! That's exactly what
> I needed but did not find.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/4894/files#r30724938.

## 

Edwin Lee, PhD
Engineer, Commercial Buildings Research Staff
National Renewable Energy Laboratory

rongpengzhang replied:
Sounds great. I will put all these minor changes in one commit.

Myoldmopar commented on doc/src/docs/InputOutputReference/01d-InputOutputReference.md:
```diff
@@ -25289,27 +25289,27 @@ This field provides the name of a Thermostat Offset Fault object that causes the
+   AlwaysOn,            !- Availability Schedule Name
+   AlwaysOne,           !- Severity Schedule Name
+   10,                  !- Reference Humidistat Offset
+   ;                    !- Related Thermostat Offset Fault Name
```
I see these idf snippets updated from the thermostat and humidistat input objects, but I don't see any actual documentation changes for the air filter objects. Those changes should be included in this branch as well in the IO and Eng Ref docs.

rongpengzhang replied:
@Myoldmopar I'm developing the Eng Ref and IO Ref for the Fault_foulingAirFilter. Probably it can be finished by tomorrow. I think we can merge this branch when it is done.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -76130,6 +76130,39 @@ FaultModel:HumidistatOffset,
        \object-list ThermostatOffsetFaults   
+	   
+FaultModel:Fouling:AirFilter,
+   \memo This object describes fault of dirty air filters
```
Might add \min-fields 6 here.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -76130,6 +76130,39 @@ FaultModel:HumidistatOffset,
+   \memo This object describes fault of dirty air filters
+   A1, \field Name
+       \required-field
+       \type alpha
```
And I know it's redundant, but you might add a \note to the Name field.  Just so the window isn't blank when a user sees this object in IDF-Editor, for example.

Myoldmopar commented:
I tell you what, @rongpengzhang.  If you are that close on docs, just push those changes up and anything else you want tomorrow or whenever you are ready.  I'll restart my major review with docs and code and everything all at once.  Tag me when everything is ready to go again.

rongpengzhang commented:
@Myoldmopar I'm still in the old working mode, when the code updates and Doc updates were separated. I'll update my working mode~

rongpengzhang committed:
Faulty Air Filters: documentation

rongpengzhang committed:
Faulty Air Filters: Codes modifications

Based on Edwin's comments

rongpengzhang commented:
@Myoldmopar 

I have finished the Documentation part, and updated the Codes part based on your suggestions. Please have a look when you are free. 

Next time I will notify you when both the Codes and Documentation parts are done.

Cheers,
Ron

rongpengzhang committed:
Merge remote-tracking branch 'origin/develop' into 83811490-Fault_DirtyAirFilter

rongpengzhang commented:
@Myoldmopar There is 1 file that didn't pass the test. I guess it is due to the update of the develop branch, so I merged the develop branch into mine.

Myoldmopar commented:
@rongpengzhang This looks pretty good.  The IDD appears to be well-formed, and the unit tests look good as well.  I have 3 comments.

### Lookup within Sim Function

One aspect I'd like you to clarify that might require a minor change.  It looks like you are looping over all the faults _every_ time ([1](https://github.com/NREL/EnergyPlus/pull/4894/files#diff-4c6021f6fbd2a312dd43625227fef114R1489), [2](https://github.com/NREL/EnergyPlus/pull/4894/files#diff-4c6021f6fbd2a312dd43625227fef114R1657), [3](https://github.com/NREL/EnergyPlus/pull/4894/files#diff-4c6021f6fbd2a312dd43625227fef114R1838)) you get into the Sim function, looking for a match.  This is really bad for performance, so I want to know if there is some reason this was intended.  If not, we need to change it.   
- At a minimum, you will want to store the fault index in the fan or whoever calls the Sim function.  If there can be multiple indices, store an array of integers.  These integers should only be initialized during GetInput and there shouldn't be any looping needed inside the simulation.  Create a simple function in the fault code that gives back the fault index by type/name.  If it can't find one, the simulation fatals out during that GetInput.  During the simulation, just pass in the index to the Sim function.  This also eliminates the need to check if NumAirFilterFaults > 0, because that will have been trapped during GetInput.  This is exactly how the CurveManager and ScheduleManager work.  There is a one-time function called `int GetCurveIndex(name)`, then you just call `Real64 CurveValue(index, values...)`.  
- In the future, when we continue flushing out the refactoring, imagine that the fan will actually hold a smart pointer to an instance of a fault.  Then instead of _ever_ looking up the index in the fault array, we will just call methods on that instance directly! :+1:

### Documentation Math

When I build the docs locally, the math isn't rendering, but it looks like you mimicked the style in the rest of the docs.  Don't make any changes right now, but maybe keep in mind to check out the math in the docs when they get built for the next iteration package.

### Compiler warnings

You have a set of compiler [warnings](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-1bf4c3587d49d8bdfad4b57f8e8604d634fb8dcb-x86_64-Linux-Ubuntu-14.04-gcc-4.8-IntegrationCoverage-Debug.html) in the faults model.  In general, I think you should have noticed those and addressed them before asking me to review this PR.  Thankfully CI reports back pretty clear with the warning message.  I cleaned these up in beef028, so no need to do anything.

Let me know if you have any questions.  It should be pretty easy to wrap up the performance issue above and get this merged in soon.

FYI @mjwitte @lgentile @hongtz68

Myoldmopar committed:
Cleanup faults default constructor; remove member constructor

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -76235,6 +76252,42 @@ FaultModel:HumidistatOffset,
+       \required-field
+       \type object-list
+       \object-list FansCVandOnOffandVAV 
+   A3, \field Fan Object Type
```
@rongpengzhang Fan Object Type field should be before Fan Object Name field to follow the pattern used elsewhere.

rongpengzhang committed:
Faulty Air Filter: address Mike and Edwin's comments

(1) IDD design for the fault model is updated
(2) Add fault index as a member of fan struct to avoid loops in
identifying fault

rongpengzhang commented:
@mjwitte  @Myoldmopar  Thanks for the advice! I've updated the codes based on your comments.

@Myoldmopar In my initial codes, I wanted to minimize the modification of existing fan codes, so I didn't add fault index as a member of fan struct. But now I know CurveManager and ScheduleManager use this approach, so I followed it. Also thank you for the modification of the Member Constructor part.

rongpengzhang commented:
@Myoldmopar I notice that 3 files failed in some tests (e.g., Win64-Windows-7-VisualStudio-12). I checked other branches (e.g. #4965), and found these files failed too. So I think that's due to the update of develop branch and can be addressed by merging develop branch into mine. 

Please let me know if you have other comments.

rongpengzhang committed:
Merge remote-tracking branch 'origin/develop' into 83811490-Fault_DirtyAirFilter

Myoldmopar committed:
Clean up documentation, update a few images

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 83811490-Fault_DirtyAirFilter

Myoldmopar commented:
@rongpengzhang I think this is ready to drop in.  I built the documentation locally and found a few issues.  For one thing, you had added images with the extension `PNG`.  While filename case doesn't matter on Windows, it does on Linux, and the images didn't show up in the documentation.  I changed those names, made a few other mods, merged develop in, and will let CI work.  After CI comes back clean, I think this will be able to merge into develop.

Subset of Checklist:
- [X] Develop merged in
- [ ] CI happy
- [X] Code review
- [X] IDD review
- [X] Documentation review
- [X] Transition - no
- [X] ExpandObjects - no

After this gets into develop, I'll push the newly rebuilt html documentation to nrel.github.io/EnergyPlus, which will include all changes to the docs for a while.

FYI @mjwitte @hongtz68

rongpengzhang commented:
@Myoldmopar Thanks for the review and modifications. I notice that you added <div> for each equation in the documentation. I will follow this style.

In the checklist, I think we can also add:
- [X] Example file review
- [X] Unit test review

Myoldmopar commented:
@rongpengzhang Your div didn't come through because GitHub is rendering the html...but I know what you are saying.  For inline math I added span and for block math I added div.  It may not have been needed, but it helps the Markdown/html/LaTeX parser.
